magic_the_gathering/phases/main.py

```diff
@@ -1,30 +1,57 @@
+from magic_the_gathering.actions.resolve_stack import ResolveTopOfStackAction
+from magic_the_gathering.game_state import GameState, ZonePosition
+from magic_the_gathering.phases.players_get_priority import PhaseWherePlayersGetPriority
+
+
+class MainPhase(PhaseWherePlayersGetPriority):
+    @staticmethod
+    def list_possible_actions(game_state: GameState) -> List[Action]:
+        possible_actions = []
+        possible_actions.extend(
+            [
+                action
+                for action in PlayLandAction.list_possible_actions(game_state)
+                if action.player_index == game_state.current_player_index
+            ]
+        )
+        possible_actions.extend(
+            [
+                action
+                for action in CastCardAction.list_possible_actions(game_state)
+                if action.player_index == game_state.current_player_index
+            ]
+        )
+        return possible_actions
 
-class MainPhase(Phase):
     def __init__(self, name: str):
         super().__init__(
             name=name,
             allows_sorcery_speed_when_stack_is_empty=True,
         )
 
     def run(self, game_state: GameState) -> GameState:
-        # TODO: Review the implementation of this method
-
-        # Current player can play a land (only 1 per turn)
-        hand_card_index = game_state.current_player.choose_land_to_play()
-        if hand_card_index is not None and not game_state.current_player_has_played_a_land_this_turn:
-            game_state.cast_card(player_index=game_state.current_player_index, hand_card_index=hand_card_index)
-            # FIXME: Here we assumes that nothing happens between casting a land and resolving it
-            game_state.resolve_top_of_the_stack()
-            game_state.current_player_has_played_a_land_this_turn = True
-
-        # Current player can cast creatures and/or sorceries
-        hand_card_indices = self.current_player.choose_creatures_to_cast()
-        for hand_card_index in hand_card_indices:
-            game_state.cast_card(player_index=game_state.current_player_index, hand_card_index=hand_card_index)
-            # FIXME: Here we assumes that nothing happens between casting a creature/sorcery and resolving it
-            game_state.resolve_top_of_the_stack()
+        while True:
+            # Current player chooses one action
+            possible_actions = PhaseWherePlayersGetPriority.list_possible_actions(game_state)
+            current_player = game_state.current_player
+            action = current_player.choose_action(
+                game_state=game_state,
+                possible_actions=possible_actions,
+            )
+            if action is None:
+                break
+            game_state = action.execute(game_state)
+
+            # TODO: Other players can respond to the action
+
+            # Resolve the stack
+            # TODO: Is it correct to resolve the whole stack at once?
```

Comment:
I don't know that it makes sense to talk about `counter actions` because the possibilities are not that different when a spell is on the stack VS not (except that sorcery speed things cannot be done even during the main phase, and that spells and abilities that targets stuff on the stack, like counterspells, can be played since there are legal targets)

Reply:
Ok I see. I guess what is missing currently is that instead of just asking the current player to do actions we should have a for loop on the players (where the current player is the first one) and each player can choose an action. With this process, it would allow other players to do actions depending on the action of the current player while keeping the implementation generic, without this concept of "counter action"